### PR TITLE
[Improve][Producer] Refactor internalSend() and resouce managment

### LINKED ISSUE #1043: [Improve][Producer]Simplify the MaxPendingMessages controlling
#### Expected behavior

1. make `pendingQueue` of dynamic size (can be expanded at runtime) or with a bigger size than `MaxPendingMessages` ;
2. `partitionProducer.canAddToQueue()` do not check `pendingQueue` any more if it can be expanded at runtime; 
3. call `partitionProducer.canAddToQueue()` in `partitionProducer.internalSendAsync` before putting the `sendRequest` into `partitionProducer.dataChan`;
4. delete  `sendRequest.blockCh`, no blocking any more;
5. treat all the messages that have been put into `partitionProducer.dataChan` as **pendingMessage**, send and flush them when the producer is closed.

Now, we won't block every time, just block when `MaxPendingMessages` is reached,  and no semantics confusion any more.

#### Actual behavior

Currently, we use a lot of measures to controll the `MaxPendingMessages` in the producer, `partitionProducer.dataChan` with `MaxPendingMessages` capacity, `partitionProducer.publishSemaphore` , `sendRequest.blockCh`, `partitionProducer.canAddToQueue()` and the complicated code in `partitionProducer.internalSend()`, these make the code difficult to understand and maintain. What is worse, is that when `DisableBlockIfQueueFull` is set to false, the `partitionProducer.internalSendAsync()` will block every time, event if the queue is NOT full, which will block the application level logic.

And, when chunking is enabled, a Message will be seperated into many chunks, correspondingly, many `pendingItem`s in the `pendingQueue`, when the `pendingQueue` is full, it doesn't mean `MaxPendingMessages` is reached from the user‘s eye view, which is semantically confused.
``` go
if !p.options.DisableBlockIfQueueFull {
  // block if queue full
	<-bc
}
```

#### Steps to reproduce

Review the code

``` go
func (p *partitionProducer) internalSendAsync(ctx context.Context, msg *ProducerMessage,
	callback func(MessageID, *ProducerMessage, error), flushImmediately bool) {
	//Register transaction operation to transaction and the transaction coordinator.
	var newCallback func(MessageID, *ProducerMessage, error)
	if msg.Transaction != nil {
		transactionImpl := (msg.Transaction).(*transaction)
		if transactionImpl.state != TxnOpen {
			p.log.WithField("state", transactionImpl.state).Error("Failed to send message" +
				" by a non-open transaction.")
			callback(nil, msg, newError(InvalidStatus, "Failed to send message by a non-open transaction."))
			return
		}

		if err := transactionImpl.registerProducerTopic(p.topic); err != nil {
			callback(nil, msg, err)
			return
		}
		if err := transactionImpl.registerSendOrAckOp(); err != nil {
			callback(nil, msg, err)
		}
		newCallback = func(id MessageID, producerMessage *ProducerMessage, err error) {
			callback(id, producerMessage, err)
			transactionImpl.endSendOrAckOp(err)
		}
	} else {
		newCallback = callback
	}
	if p.getProducerState() != producerReady {
		// Producer is closing
		newCallback(nil, msg, errProducerClosed)
		return
	}

	// bc only works when DisableBlockIfQueueFull is false
	bc := make(chan struct{})

	// callbackOnce make sure the callback is only invoked once in chunking
	callbackOnce := &sync.Once{}
	var txn *transaction
	if msg.Transaction != nil {
		txn = (msg.Transaction).(*transaction)
	}
	sr := &sendRequest{
		ctx:              ctx,
		msg:              msg,
		callback:         newCallback,
		callbackOnce:     callbackOnce,
		flushImmediately: flushImmediately,
		publishTime:      time.Now(),
		blockCh:          bc,
		closeBlockChOnce: &sync.Once{},
		transaction:      txn,
	}
	p.options.Interceptors.BeforeSend(p, msg)

	p.dataChan <- sr

	if !p.options.DisableBlockIfQueueFull {
		// block if queue full
		<-bc
	}
}

func (p *partitionProducer) canAddToQueue(sr *sendRequest, uncompressedPayloadSize int64) bool {
	if p.options.DisableBlockIfQueueFull {
		if !p.publishSemaphore.TryAcquire() {
			if sr.callback != nil {
				sr.callback(nil, sr.msg, errSendQueueIsFull)
			}
			return false
		}
		if !p.client.memLimit.TryReserveMemory(uncompressedPayloadSize) {
			p.publishSemaphore.Release()
			if sr.callback != nil {
				sr.callback(nil, sr.msg, errMemoryBufferIsFull)
			}
			return false
		}

	} else {
		if !p.publishSemaphore.Acquire(sr.ctx) {
			sr.callback(nil, sr.msg, errContextExpired)
			return false
		}
		if !p.client.memLimit.ReserveMemory(sr.ctx, uncompressedPayloadSize) {
			p.publishSemaphore.Release()
			sr.callback(nil, sr.msg, errContextExpired)
			return false
		}
	}
	p.metrics.MessagesPending.Inc()
	p.metrics.BytesPending.Add(float64(len(sr.msg.Payload)))
	return true
}

func (p *partitionProducer) internalSend(request *sendRequest) {
	p.log.Debug("Received send request: ", *request.msg)

	msg := request.msg

	// read payload from message
	uncompressedPayload := msg.Payload
	uncompressedPayloadSize := int64(len(uncompressedPayload))

	var schemaPayload []byte
	var err error
	if msg.Value != nil && msg.Payload != nil {
		p.log.Error("Can not set Value and Payload both")
		request.callback(nil, request.msg, errors.New("can not set Value and Payload both"))
		return
	}

	// The block chan must be closed when returned with exception
	defer request.stopBlock()
	if !p.canAddToQueue(request, uncompressedPayloadSize) {
		return
	}

	if p.options.DisableMultiSchema {
		if msg.Schema != nil && p.options.Schema != nil &&
			msg.Schema.GetSchemaInfo().hash() != p.options.Schema.GetSchemaInfo().hash() {
			p.releaseSemaphoreAndMem(uncompressedPayloadSize)
			request.callback(nil, request.msg, fmt.Errorf("msg schema can not match with producer schema"))
			p.log.WithError(err).Errorf("The producer %s of the topic %s is disabled the `MultiSchema`", p.producerName, p.topic)
			return
		}
	}
	var schema Schema
	var schemaVersion []byte
	if msg.Schema != nil {
		schema = msg.Schema
	} else if p.options.Schema != nil {
		schema = p.options.Schema
	}
	if msg.Value != nil {
		// payload and schema are mutually exclusive
		// try to get payload from schema value only if payload is not set
		if uncompressedPayload == nil && schema != nil {
			schemaPayload, err = schema.Encode(msg.Value)
			if err != nil {
				p.releaseSemaphoreAndMem(uncompressedPayloadSize)
				request.callback(nil, request.msg, newError(SchemaFailure, err.Error()))
				p.log.WithError(err).Errorf("Schema encode message failed %s", msg.Value)
				return
			}
		}
	}
	if uncompressedPayload == nil {
		uncompressedPayload = schemaPayload
	}

	if schema != nil {
		schemaVersion = p.schemaCache.Get(schema.GetSchemaInfo())
		if schemaVersion == nil {
			schemaVersion, err = p.getOrCreateSchema(schema.GetSchemaInfo())
			if err != nil {
				p.releaseSemaphoreAndMem(uncompressedPayloadSize)
				p.log.WithError(err).Error("get schema version fail")
				request.callback(nil, request.msg, fmt.Errorf("get schema version fail, err: %w", err))
				return
			}
			p.schemaCache.Put(schema.GetSchemaInfo(), schemaVersion)
		}
	}

	uncompressedSize := len(uncompressedPayload)

	deliverAt := msg.DeliverAt
	if msg.DeliverAfter.Nanoseconds() > 0 {
		deliverAt = time.Now().Add(msg.DeliverAfter)
	}

	mm := p.genMetadata(msg, uncompressedSize, deliverAt)

	// set default ReplicationClusters when DisableReplication
	if msg.DisableReplication {
		msg.ReplicationClusters = []string{"__local__"}
	}

	sendAsBatch := !p.options.DisableBatching &&
		msg.ReplicationClusters == nil &&
		deliverAt.UnixNano() < 0

	// Once the batching is enabled, it can close blockCh early to make block finish
	if sendAsBatch {
		request.stopBlock()
	} else {
		// update sequence id for metadata, make the size of msgMetadata more accurate
		// batch sending will update sequence ID in the BatchBuilder
		p.updateMetadataSeqID(mm, msg)
	}

	maxMessageSize := int(p._getConn().GetMaxMessageSize())

	// compress payload if not batching
	var compressedPayload []byte
	var compressedSize int
	var checkSize int
	if !sendAsBatch {
		compressedPayload = p.compressionProvider.Compress(nil, uncompressedPayload)
		compressedSize = len(compressedPayload)
		checkSize = compressedSize

		// set the compress type in msgMetaData
		compressionType := pb.CompressionType(p.options.CompressionType)
		if compressionType != pb.CompressionType_NONE {
			mm.Compression = &compressionType
		}
	} else {
		// final check for batching message is in serializeMessage
		// this is a double check
		checkSize = uncompressedSize
	}

	// if msg is too large and chunking is disabled
	if checkSize > maxMessageSize && !p.options.EnableChunking {
		p.releaseSemaphoreAndMem(uncompressedPayloadSize)
		request.callback(nil, request.msg, errMessageTooLarge)
		p.log.WithError(errMessageTooLarge).
			WithField("size", checkSize).
			WithField("properties", msg.Properties).
			Errorf("MaxMessageSize %d", maxMessageSize)
		p.metrics.PublishErrorsMsgTooLarge.Inc()
		return
	}

	var totalChunks int
	// max chunk payload size
	var payloadChunkSize int
	if sendAsBatch || !p.options.EnableChunking {
		totalChunks = 1
		payloadChunkSize = int(p._getConn().GetMaxMessageSize())
	} else {
		payloadChunkSize = int(p._getConn().GetMaxMessageSize()) - proto.Size(mm)
		if payloadChunkSize <= 0 {
			p.releaseSemaphoreAndMem(uncompressedPayloadSize)
			request.callback(nil, msg, errMetaTooLarge)
			p.log.WithError(errMetaTooLarge).
				WithField("metadata size", proto.Size(mm)).
				WithField("properties", msg.Properties).
				Errorf("MaxMessageSize %d", int(p._getConn().GetMaxMessageSize()))
			p.metrics.PublishErrorsMsgTooLarge.Inc()
			return
		}
		// set ChunkMaxMessageSize
		if p.options.ChunkMaxMessageSize != 0 {
			payloadChunkSize = int(math.Min(float64(payloadChunkSize), float64(p.options.ChunkMaxMessageSize)))
		}
		totalChunks = int(math.Max(1, math.Ceil(float64(compressedSize)/float64(payloadChunkSize))))
	}

	// set total chunks to send request
	request.totalChunks = totalChunks

	if !sendAsBatch {
		if totalChunks > 1 {
			var lhs, rhs int
			uuid := fmt.Sprintf("%s-%s", p.producerName, strconv.FormatUint(*mm.SequenceId, 10))
			mm.Uuid = proto.String(uuid)
			mm.NumChunksFromMsg = proto.Int32(int32(totalChunks))
			mm.TotalChunkMsgSize = proto.Int32(int32(compressedSize))
			cr := newChunkRecorder()
			for chunkID := 0; chunkID < totalChunks; chunkID++ {
				lhs = chunkID * payloadChunkSize
				if rhs = lhs + payloadChunkSize; rhs > compressedSize {
					rhs = compressedSize
				}
				// update chunk id
				mm.ChunkId = proto.Int32(int32(chunkID))
				nsr := &sendRequest{
					ctx:              request.ctx,
					msg:              request.msg,
					callback:         request.callback,
					callbackOnce:     request.callbackOnce,
					publishTime:      request.publishTime,
					blockCh:          request.blockCh,
					closeBlockChOnce: request.closeBlockChOnce,
					totalChunks:      totalChunks,
					chunkID:          chunkID,
					uuid:             uuid,
					chunkRecorder:    cr,
					transaction:      request.transaction,
				}
				// the permit of first chunk has acquired
				if chunkID != 0 && !p.canAddToQueue(nsr, 0) {
					p.releaseSemaphoreAndMem(uncompressedPayloadSize - int64(rhs))
					return
				}
				p.internalSingleSend(mm, compressedPayload[lhs:rhs], nsr, uint32(maxMessageSize))
			}
			// close the blockCh when all the chunks acquired permits
			request.stopBlock()
		} else {
			// close the blockCh when totalChunks is 1 (it has acquired permits)
			request.stopBlock()
			p.internalSingleSend(mm, compressedPayload, request, uint32(maxMessageSize))
		}
	} else {
		smm := p.genSingleMessageMetadataInBatch(msg, uncompressedSize)
		multiSchemaEnabled := !p.options.DisableMultiSchema
		added := addRequestToBatch(smm, p, uncompressedPayload, request, msg, deliverAt, schemaVersion,
			multiSchemaEnabled)
		if !added {
			// The current batch is full. flush it and retry

			p.internalFlushCurrentBatch()

			// after flushing try again to add the current payload
			if ok := addRequestToBatch(smm, p, uncompressedPayload, request, msg, deliverAt, schemaVersion,
				multiSchemaEnabled); !ok {
				p.releaseSemaphoreAndMem(uncompressedPayloadSize)
				request.callback(nil, request.msg, errFailAddToBatch)
				p.log.WithField("size", uncompressedSize).
					WithField("properties", msg.Properties).
					Error("unable to add message to batch")
				return
			}
		}
		if request.flushImmediately {

			p.internalFlushCurrentBatch()

		}
	}
}
```

#### System configuration
**Pulsar version**: x.y
@merlimat @RobertIndie @wolfstudy @Gleiphir2769

### PR DESCRIPTION
**Pull request by gunli**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #1043

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #1043, #1055 #1059 #1060 #1067, #1068

### Motivation
1. As discussed in #1055, we need to calculate how many pending items and how many memory are required before appending the `sendRequest` to the `dataChan`, but currently, we do schema-encoding/compressing in `internalSend()`, this may lead to inaccurate memory limit cotrolling, and as described in #1043, make the code complicated and difficult to maintain, we need to simplify the send logic;
2. In JAVA client, schema-encoding/compressing are done in application thread, it better to make it work like JAVA client; 
3. As discussed in #1055 and described in #1043, resource(memory and semaphore) acquiring/releasing logic are written across the whole file, we need to simplify the resource management logic, encapsulate them into functions, call them when necessary, make it 'Low Coupling, High Cohesion';
4. As discussed in #1060, transaction is not correctly ended for chunked message, it better to encapsulate the transaction ending logic into one func which will be called when sendRequest is done.

### Modifications

1. Move shema encoding from `internalSend()` to `internalSendAsync()`;
2. Move compressing from `internalSend()` to `internalSendAsync()`;
5. Calculate total chunks before entering the dataChan;
6. Reserve required semaphore and memory before entering the dataChan;
7. `sendRequest` store the semaphore and memory it holds; 
8. Encapsualte relative code blocks into individual funcs to make the skeleton of  `internalSendAsync()` clearer;
9. Add `sendRequest.done()` to release the resources it holds;
10. When a `sendRequest` is done, call `sendRequest.done()` ;
11. In `sendRequest.done()` run callback, update metrics, end transaction, run interceptors callback...

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change is already covered by existing tests, such as *(please describe tests)*.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)

### Documentation

  - Does this pull request introduce a new feature? no)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by gunli** [comment]
@Gleiphir2769 @RobertIndie @graysonzeng @shibd Would you please review this PR ?

**Review by Gleiphir2769 — COMMENTED** [review]
Hi @gunli, I leaved some comments.

File: pulsar/producer_partition.go [hunk 1]
```diff
@@ -1121,78 +958,301 @@ func (p *partitionProducer) SendAsync(ctx context.Context, msg *ProducerMessage,
...
+	}
+	if err := p.reserveMem(sr); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *partitionProducer) internalSendAsync(ctx context.Context, msg *ProducerMessage,
+	callback func(MessageID, *ProducerMessage, error), flushImmediately bool) {
+	err := p.validateMsg(msg)
```
[the remark this answers is not in the document]
Comment (gunli) [reply pulsar/producer_partition.go:1191]: inline will make internalSendAsync a BIG func, about 200 lines, it hard to read, spilt into small funcs will be more clear.

File: pulsar/producer_partition.go [hunk 2]
```diff
@@ -1121,78 +958,301 @@ func (p *partitionProducer) SendAsync(ctx context.Context, msg *ProducerMessage,
...
-			return
+		p.schemaCache.Put(schema.GetSchemaInfo(), schemaVersion)
+	}
+
+	sr.schema = schema
+	sr.schemaVersion = schemaVersion
+	return nil
+}
+
+func (p *partitionProducer) updateUncompressPayload(sr *sendRequest) error {
```
[the remark this answers is not in the document]
Comment (gunli) [reply pulsar/producer_partition.go:1010]: OK, I will rename it.

File: pulsar/producer_partition.go [hunk 3]
```diff
@@ -1121,78 +958,301 @@ func (p *partitionProducer) SendAsync(ctx context.Context, msg *ProducerMessage,
...
+		// set the compress type in msgMetaData
+		compressionType := pb.CompressionType(p.options.CompressionType)
+		if compressionType != pb.CompressionType_NONE {
+			sr.mm.Compression = &compressionType
+		}
+
+		checkSize = int64(sr.compressedSize)
+	}
+
+	sr.maxMessageSize = int32(int64(p._getConn().GetMaxMessageSize()))
```
[the remark this answers is not in the document]
Comment (gunli) [reply pulsar/producer_partition.go:1076]: MaxMessageSize is cached in the conn when a conn is ready, you can check the code of conneciton and connectionPool, this is OK.

[the remark this answers is not in the document]
Comment (Gleiphir2769) [reply pulsar/producer_partition.go:1076]: > MaxMessageSize is cached in the conn when a conn is ready, you can check the code of connection and connectionPool, this is OK.

You're right. This point is good to me.

> Actually, p.getOrCreateSchema() will trigger a block RPC call. What make SendAsync not a real asyn func is the fixed length PendingQueue

But the size of `PendingQueue` is decided by user. And if the size of `dataChan` is set as `MaxPendingMessages`, it will not be the async limit.

> I don't think memLimit and fixed length queue is neccessary in a language with GC like JAVA and Go

This is not only related to memory, but also related to broker. Here is the interface description of Java client.
>Set the max size of the queue holding the messages pending to receive an acknowledgment from the broker

I think we should find a way to solve the `p.getOrCreateSchema()` problem. What do you think?

File: pulsar/producer_partition.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gleiphir2769) [reply pulsar/producer_partition.go:1191]: Sorry, maybe my description is not clear. Could we make L1191-L1192 as a `One Liner if...else Statements`.

File: pulsar/producer_partition.go [hunk 4]
```diff
@@ -1121,78 +958,301 @@ func (p *partitionProducer) SendAsync(ctx context.Context, msg *ProducerMessage,
...
 		publishTime:      time.Now(),
-		blockCh:          bc,
-		closeBlockChOnce: &sync.Once{},
-		transaction:      txn,
+		flushImmediately: flushImmediately,
 	}
-	p.options.Interceptors.BeforeSend(p, msg)
 
-	p.dataChan <- sr
+	err = p.updateSchema(sr)
```
[the remark this answers is not in the document]
Comment (Gleiphir2769) [reply pulsar/producer_partition.go:1217]: Same as above

File: pulsar/producer_partition.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gunli) [reply pulsar/producer_partition.go:1077]: > But the size of `PendingQueue` is decided by user. And if the size of `dataChan` is set as `MaxPendingMessages`, it will not be the async limit.

dataChan and semaphore can work together to make it async, when there is enough semaphore, add to the queue, or wait until the response back from the broker and release a semaphore.



> I think we should find a way to solve the `p.getOrCreateSchema()` problem.

Since the schema can be set by the user at runtime, we can only solve it by moving the blocking logic to `internalSend()',  but at the same time, we have to reserve memory and peding queue spaces before adding a message to dataChan, so if we want to solve it, we must drop memLimit and the fixed length pending queue.

File: pulsar/producer_partition.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gunli) [reply pulsar/producer_partition.go:1192]: Hmm, I think that is OK. inlining can reduce the code line number, non-inlining are better for debugging.

File: pulsar/producer_partition.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gleiphir2769) [reply pulsar/producer_partition.go:1077]: > we must drop memLimit and the fixed length pending queue

Sorry, I am not get the point why we should drop the memLimit? It's a useful feature for the users who is lack of resources. 
And I don't think the `fixed length pending queue` will become a problem to `SendAsync`. Why should we make it flexible?

[the remark this answers is not in the document]
Comment (gunli) [reply pulsar/producer_partition.go:1077]: > I support the idea which reserve resource before `internalSend`. And if we reserve the semaphore firstly, the block of `dataChan` will not happen if it's made with capacity `MaxPendingMessages`.

We do not block on dataChan, just block on semaphore, we just treat dataChan as a channel between the main goroutine(user's goroutine) and the partitionProducer's goroutine(IO goroutine), semaphore represents the available resource(pendingItem), When aquire semaphore succeed, we can add it to dataChan, otherwise, block until one semaphore(one message has been done) .



> Sorry, I am not get the point why we should drop the memLimit? It's a useful feature for the users who is lack of resources.
> And I don't think the `fixed length pending queue` will become a problem to `SendAsync`. Why should we make it flexible?

Because we limit the memory and pending queue, we have to reserve memory and pending queue before adding a message to dataChan, which make we have to do schema encoding and compressing first, or can not know how much memory and how many pending item count we need, these are bloking logic. When these blocking jobs are done in the user's goroutine, the block the user's logic, which make it a non-async method.

**Comment by gunli** [comment]
@RobertIndie Would you please take a look at this PR, it pending for a long time.

**Comment by gunli** [comment]
@tisonkun @RobertIndie I think there is a bug in `reserveSemaphore()`, I will update it now.

**Comment by gunli** [comment]
> @tisonkun @RobertIndie I think there is a bug in `reserveSemaphore()`, I will update it now.

done.

**Comment by RobertIndie** [comment]
I noticed that you did a lot of refactoring work in this PR, including changes to the critical path for publishing messages. I’m concerned that these changes may impact publishing performance.

While there are many modifications, it’s difficult for the reviewer to see the relationship between these modifications and the bug fixes. Could you provide more detail in the PR description, particularly regarding why you’re making these changes to fix bugs?

I recommend separating the refactoring work from the bug fixes. Otherwise, it will be challenging for us to cherry-pick the bug fixes to other release branches.

**Comment by gunli** [comment]
> I noticed that you did a lot of refactoring work in this PR, including changes to the critical path for publishing messages. I’m concerned that these changes may impact publishing performance.
> 
> While there are many modifications, it’s difficult for the reviewer to see the relationship between these modifications and the bug fixes. Could you provide more detail in the PR description, particularly regarding why you’re making these changes to fix bugs?
> 
> I recommend separating the refactoring work from the bug fixes. Otherwise, it will be challenging for us to cherry-pick the bug fixes to other release branches.

@RobertIndie The details we have disccussed in #1059 #1043 #1060, the bug fix just fixed the bug in the PR itself, not the exist code.
As we disscussed in #1059, we have to move some logic from internalSend() to internalSendAsync(), these logic work together, so I can't seperate them into multiple PRs.

**Comment by RobertIndie** [comment]
@gunli Could you summarize them and add the detailed explanation to the PR description?  They are very important context for this PR. A well-written PR description not only helps with review but also enables other developers to learn about the context of this PR.

**Comment by gunli** [comment]
> @gunli Could you summarize them and add the detailed explanation to the PR description? They are very important context for this PR. A well-written PR description not only helps with review but also enables other developers to learn about the context of this PR.

@RobertIndie I have updated the motivation.

**Comment by Gleiphir2769** [comment]
Hi @gunli . Looks like it does not pass the CI test now. You can check the CI log an fix it. Seems like it is related of Chunking.

image

**Comment by gunli** [comment]
> 

@Gleiphir2769 Thank you, I am busy these days, I will fix it ASAP.

**Comment by gunli** [comment]
> > 
> 
> @Gleiphir2769 Thank you, I am busy these days, I will fix it ASAP.

@Gleiphir2769 I have fix the test case, and it can run PASS now, but since only the last chunk will trigger the callback, I commented the wait logic in `sendSingleChunk()`, I am not sure whether it is right to do that, please help to review it.

File: pulsar/message_chunking_test.go [hunk 5]
```diff
@@ -548,30 +556,57 @@ func createTestMessagePayload(size int) []byte {
 }
 
 //nolint:all
-func sendSingleChunk(p Producer, uuid string, chunkID int, totalChunks int) {
+func sendSingleChunk(p Producer, uuid string, chunkID int, totalChunks int, wholePayload string, callbackOnce *sync.Once, cr *chunkRecorder) {
```
Comment (Gleiphir2769) [inline pulsar/message_chunking_test.go:559]: Should we add the parameters`wholePayload ` `callbackOnce ` and `chunkRecorder `? I think it's more appropriate to make them inside the`sendSingleChunk `.

**Review by Gleiphir2769 — COMMENTED** [review]
Hi @gunli. I leaved some comments about `sendSingleChunk `.

File: pulsar/message_chunking_test.go (hunk 5 above)
Comment (gunli) [reply pulsar/message_chunking_test.go:559]: Hmm, in the current implemention, all the chunking sendRequest should share the whole message payload, callbackOnce, chunkRecorder, see `internalSend()` and `addRequestToBatch()`, I add these params just want to make the send procedure work correctly, I think it better to keep them.

**Comment by gunli** [comment]
@RobertIndie @tisonkun Would you please review this PR again?

**Comment by gunli** [comment]
[message_chunking_test-ac9c1a6399336461d2d3ce1cdd31cac6debd5ed5.txt](https://github.com/apache/pulsar-client-go/files/12977288/message_chunking_test-ac9c1a6399336461d2d3ce1cdd31cac6debd5ed5.txt)
[message_chunking_test-PR.txt](https://github.com/apache/pulsar-client-go/files/12977289/message_chunking_test-PR.txt)
[producer_partition-ac9c1a6399336461d2d3ce1cdd31cac6debd5ed5.txt](https://github.com/apache/pulsar-client-go/files/12977290/producer_partition-ac9c1a6399336461d2d3ce1cdd31cac6debd5ed5.txt)
[producer_partition-PR.txt](https://github.com/apache/pulsar-client-go/files/12977291/producer_partition-PR.txt)
@RobertIndie @tisonkun I have uploaded the files that have been changed, would you please review this PR with some compareing tool line BeyondCompare?

**Comment by gunli** [comment]
> This PR looks complex, could you split this PR?

@nodece OK, I will do it today.
